compiler/definitions/ir/resource.py

```diff
@@ -56,3 +57,49 @@ def __eq__(self, other):
 class EphemeralResource(Resource):
     def __init__(self):
         self.uri = None
+
+class RemoteFileResource(Resource):
+    def __init__(self):
+        raise NotImplementedError("RemoteFileResource is an interface")
+
+    def __str__(self):
+        raise NotImplementedError("RemoteFileResource is an interface")
+
+    def is_available_on(self, host):
+        raise NotImplementedError("RemoteFileResource is an interface")
+
+    def _normalize_addr(self, addr):
+        """
+        Helper, removes port number if supplied in address and normalizes host address
+        Example1: datanode1 -> 172.18.0.3
+        Example2: 172.18.0.3:5555 -> 172.18.0.3
+        """
+        host = addr.rsplit(":", 1)[0]
+        normalized_host = socket.gethostbyaddr(host)[2][0]
+        return normalized_host
+
+class HDFSFileResource(RemoteFileResource):
+    ## The uri is the path of the file.
+    def __init__(self, uri, resource_hosts):
+        """
+        Params:
+            uri: Usually the path to the file
+            resource_hosts: the addresses of all the machines containing 
+            the resource.
+        """
+        self.uri = uri
+        self.hosts = list(map(self._normalize_addr, resource_hosts))
+
+    def __eq__(self, other):
+        if isinstance(other, HDFSFileResource):
+            return self.uri == other.uri
+        return False
+
+    def is_available_on(self, host):
+        return host in self.hosts
+
+    def __repr__(self):
+        return f'hdfs://{str(self.uri)}'
```

Comment:
I don't think you need to add `str` inside format strings. For example,  `f'{1}` should work and call __repr__ for the value automatically.